dbt/adapters/redshift/relation_configs/dist.py

```diff
@@ -0,0 +1,164 @@
+from dataclasses import dataclass
+from typing import Optional, Set
+
+import agate
+from dbt.adapters.relation_configs import (
+    RelationConfigChange,
+    RelationConfigChangeAction,
+    RelationConfigValidationMixin,
+    RelationConfigValidationRule,
+)
+from dbt.contracts.graph.nodes import ModelNode
+from dbt.dataclass_schema import StrEnum
+from dbt.exceptions import DbtRuntimeError
+
+from dbt.adapters.redshift.relation_configs.base import RedshiftRelationConfigBase
+
+
+class RedshiftDistStyle(StrEnum):
+    auto = "auto"
+    even = "even"
+    all = "all"
+    key = "key"
+
+    @classmethod
+    def default(cls) -> "RedshiftDistStyle":
+        return cls.auto
+
+
+@dataclass(frozen=True, eq=True, unsafe_hash=True)
+class RedshiftDistConfig(RedshiftRelationConfigBase, RelationConfigValidationMixin):
+    """
+    This config fallows the specs found here:
+    https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_NEW.html
+
+    The following parameters are configurable by dbt:
+    - diststyle: the type of data distribution style to use on the table/materialized view
+    - distkey: the column to use for the dist key if `dist_style` is `key`
+    """
+
+    diststyle: Optional[RedshiftDistStyle] = RedshiftDistStyle.default()
+    distkey: Optional[str] = None
+
+    @property
+    def validation_rules(self) -> Set[RelationConfigValidationRule]:
+        # index rules get run by default with the mixin
+        return {
+            RelationConfigValidationRule(
+                validation_check=not (
+                    self.diststyle == RedshiftDistStyle.key and self.distkey is None
+                ),
+                validation_error=DbtRuntimeError(
+                    "A `RedshiftDistConfig` that specifies a `diststyle` of `key` must provide a value for `distkey`."
+                ),
+            ),
+            RelationConfigValidationRule(
+                validation_check=not (
+                    self.diststyle
+                    in (RedshiftDistStyle.auto, RedshiftDistStyle.even, RedshiftDistStyle.all)
+                    and self.distkey is not None
+                ),
+                validation_error=DbtRuntimeError(
+                    "A `RedshiftDistConfig` that specifies a `distkey` must be of `diststyle` `key`."
+                ),
+            ),
+        }
+
+    @classmethod
+    def from_dict(cls, config_dict) -> "RedshiftDistConfig":
```

Comment:
I'm mirroring frameworks like `mashumaro` since I suspect we might replace this with something like that anyway. Even if we don't, it should evoke those types of feels.